File: reb_examples/reb_19_5_1/reb_19_5_1_diff.m
function reb_19_5_1_diff()
%reb_16_5_1 Reaction Engineering Basics Example 19.5.1

    % given and known 
    V = 1.0; % L
    R = 8.314E-3; % kJ /mol /K

    % function that performs the calculations
    function perform_the_calculations()
        % read the experimental data
        data_file = '../reb_19_5_1_data.csv';
        data_table = readtable(data_file, 'VariableNamingRule'...
            , 'preserve');
        data = table2array(data_table(:,2:end));

        % get the temperatures of the data blocks
        block_temperatures = unique(data(:,1));
        n_blocks = size(block_temperatures,1);

        % create vectors to store the fitting results
        k_fit = nan(n_blocks,1);
        kll_fit = nan(n_blocks,1);
        kul_fit = nan(n_blocks,1);
        rsq_fit = nan(n_blocks,1);

        % process the data blocks
        for iBlock = 1:n_blocks
            % create the block
            idx = table2array(data_table(:,2) == block_temperatures(iBlock));
            data_block = data(idx,:);

            % extract the data as arrays
            CA0 = data_block(:,2);
            tf = data_block(:,3);
            CAf = data_block(:,4);
            nData = length(tf);

            % calculate x
            x = CAf*V;

            % allocate storage for y
            y = nan(length(tf),1);

            % calculate y
            for i = 1:length(tf)
                if tf(i) == 5.0
                    x_minus = CA0(i)*V;
                    t_minus = 0;
                else
                    x_minus = x(i-1);
                    t_minus = tf(i-1);
                end
                y(i) = (x_minus - x(i))/(tf(i) - t_minus);
            end

            % fit y = m*x to the data
            beta = inv(x.'*x)*x.'*y;

            % calculate the model-predicted y
            y_model = x*beta;

            % calculate the 95% confidence interval
            nParams = length(beta);
            eps = y - x*beta;
            varEps = 1/(nData - nParams)*(eps.'*eps);
            covBeta = varEps*inv(x.'*x);
            tCrit = tinv(0.975,(nData - nParams));
            betaCI = sqrt(covBeta)*tCrit;

            % calculate R-squared
            ybar = sum(y)/nData;
            r_squared = (sum((y-ybar).^2) - sum((y-y_model).^2))/sum((y-ybar).^2);

            % save the fitting results 
            k_fit(iBlock) = beta;
            kll_fit(iBlock) = beta - betaCI;
            kul_fit(iBlock) = beta + betaCI;
            rsq_fit(iBlock) = r_squared;

            % create, show, and save a model plot
            figure
            hold on
            plot(x,y_model,'r','LineWidth',2)
            plot(x, y,'ok','MarkerSize',10,'LineWidth',2)
            hold off
            set(gca, 'FontSize', 14);
            xlabel('x','FontSize', 14)
            ylabel('y','FontSize', 14)
            fname = strcat('reb_19_5_1_model_'...
                , int2str(block_temperatures(iBlock)),'.png');
            saveas(gcf,fname)
        end

        % tabulate, show, and save the results
        results_file ="reb_19_5_1_diff_params.csv";
        results_table = table(block_temperatures, k_fit, kll_fit...
            , kul_fit, rsq_fit);
        results_table.Properties.VariableNames = ["T", "k", "k_ll"...
            , "k_ul", "R_sq"];
        disp(results_table)
        writetable(results_table,results_file);

        % fit the Arrhenius expression to the k vs. T data
        T = block_temperatures + 273.15;
        [k0, k0_ci, E, E_ci, r_squared] = Arrhenius_parameters(k_fit...
            ,T,R);

        % tabulate, show, and save the results
        item = ["k0"; "k0_lower_limit"; "k0_upper_limit"; "E"; 
            "E_lower_limit"; "E_upper_limit"; "R_squared"];
        value = [k0; k0_ci(1); k0_ci(2); E; E_ci(1); E_ci(2);
             r_squared];
        units = ["min^-1^"; "min^-1^"; "min^-1^"; "kJ mol^-1^";
            "kJ mol^-1^"; "kJ mol^-1^"; ""];
        Arrhenius_table = table(item, value, units);
        disp(Arrhenius_table)
        writetable(Arrhenius_table,'reb_19_5_1_Arrhenius_diff.csv')

        % create, show and save an Arrhenius plot
        k_pred = k0*exp(-E/R./T);
        x = 1./T;
        figure
        semilogy(x,k_fit,'ok',x,k_pred,'r','MarkerSize',10 ...
            ,'LineWidth',2)
        set(gca, 'FontSize', 14);
        xlabel('1/T (/K)','FontSize', 14)
        ylabel('k (/min)','FontSize', 14)
        saveas(gcf,'reb_19_5_1_Arrhenius_diff.png')
    end

    % perform the calculations
    perform_the_calculations()
end

File: reb_examples/reb_19_5_1/reb_19_5_1_diff.py
"""Calculations for Reaction Engineering Basics Example 19.5.1 using differential analysis"""

# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
from score_utils import Arrhenius_parameters

# given and known constants
V = 1.0 # L
R = 8.314E-3 # kJ/mol/K

# function that performs the calculations
def perform_the_calculations():
    # read the data from the .csv file
    df = pd.read_csv('reb_19_5_1/reb_19_5_1_data.csv')
        # columns: Experiment, T, CA0, tf, CAf
    
    # get the temperatures of the data blocks
    block_temperatures = df['T'].unique()

    # create a dataframe for the fitting results from the same-temperature blocks
    parameter_estimates = pd.DataFrame(columns=['T', 'k', 'k_ll', 'k_ul', 'R_sq'])

    # process the data blocks
    for T in block_temperatures:
        # create the block
        block = df[df['T'] == T]

        # extract the data as arrays
        T_expts = block['T'].to_numpy()
        CA0 = block['CA0'].to_numpy()
        tf = block['tf'].to_numpy()
        CAf = block['CAf'].to_numpy()

        # calculate x
        x = block['CAf'].to_numpy()*V

        # allocate storage for y
        y = np.zeros(len(block.index))

        # calculate y
        for i in range(0,len(x)):
            if tf[i] == 5.0:
                x_minus = CA0[i]*V
                t_minus = 0
            else:
                x_minus = x[i-1]
                t_minus = tf[i-1]
            y[i] = (x_minus - x[i])/(tf[i] - t_minus)
        
        # fit y = mx to the data
        model = sm.OLS(y,x)
        res = model.fit()
        m = res.params[0]
        m_ci = res.conf_int(alpha=0.05)

        # add the fitting results to the dataframe
        parameter_estimates.loc[len(parameter_estimates)] = [T, m, m_ci[0,0]
                    , m_ci[0,1], res.rsquared]
    
        # calculate the model-predicted y
        y_model = m*x
        
        # create, show, and save a model plot
        plt.figure() 
        plt.plot(x, y_model, color = 'r', ls='-' , label = 'model')
        plt.plot(x, y, color = 'k', marker='o', ls='', label = 'experiment')
        plt.xlabel("x")
        plt.ylabel("y")
        T_as_text = format(T,'.0f')
        plt.legend(title='T = ' + T_as_text + '°C')
        plt.tight_layout()
        f_name = 'reb_19_5_1_model_' + T_as_text + '.png'
        plt.savefig('reb_19_5_1/python/' + f_name)
        plt.show()

    # show and save the fitting results
    print('\nParameter Estimates:\n')
    print(parameter_estimates)
    parameter_estimates.to_csv(
        'reb_19_5_1/python/reb_19_5_1_diff_params.csv', index=False)
    
    # fit the Arrhenius expression to the k vs T data
    T_block = parameter_estimates['T'].to_numpy() + 273.15
    k = parameter_estimates['k'].to_numpy()
    # fit the Arrhenius expression to the data
    k0, k0_ci, E, E_ci, r_squared = Arrhenius_parameters(k,T_block,R)

    # report the results
    print(' ')
    print(f'k0: {k0:.3g} /min, 95% CI [{k0_ci[0]:.3g}, {k0_ci[1]:.3g}]')
    print(f'E: {E:.3g} kJ/mol, 95% CI [{E_ci[0]:.3g}, {E_ci[1]:.3g}]')
    print(f'R-squared: {r_squared:.3g}')
    print(' ')

    # save the results to a .csv file
    data = [['k0', f'{k0:.3g}', 'min^-1^'],
        ['k0_lower_limit', f'{k0_ci[0]:.3g}', 'min^-1^'],
        ['k0_upper_limit', f'{k0_ci[1]:.3g}', 'min^-1^'],
        ['E', f'{E:.3g}', 'kJ mol^-1^'],
        ['E_lower_limit', f'{E_ci[0]:.3g}', 'kJ mol^-1^'],
        ['E_upper_limit', f'{E_ci[1]:.3g}', 'kJ mol^-1^'],
        ['R_squared', f'{r_squared:.3g}', '']]
    result = pd.DataFrame(data, columns=['item','value','units'])
    result.to_csv("reb_19_5_1/python/reb_19_5_1_Arrhenius_diff.csv", 
                index=False)

    # create, show, and save an Arrhenius plot
    y_pred = k0*np.exp(-E/R/T_block)
    plt.figure()
    plt.semilogy(1/T_block,k,color='k',marker='o', ls='none')
    plt.semilogy(1/T_block,y_pred,color='r')
    plt.xlabel('T$^{-1}$ (K$^{-1}$)')
    plt.ylabel('k (min$^{-1}$)')
    plt.xticks(rotation=25)
    plt.tight_layout()
    plt.savefig('reb_19_5_1/python/reb_19_5_1_Arrhenius_diff.png')
    plt.show()

if __name__=="__main__":
    perform_the_calculations()